thingsvision/core/extraction/mixin.py

```diff
@@ -0,0 +1,164 @@
+import warnings
+from dataclasses import dataclass, field
+from typing import Any, Iterator, Tuple
+
+import numpy as np
+import tensorflow as tf
+import torch
+import torchvision
+from tensorflow import keras
+from tensorflow.keras import layers
+from torchvision import transforms as T
+from tqdm import tqdm
+
+Tensor = torch.Tensor
+Array = np.ndarray
+
+
+@dataclass
+class PyTorchMixin:
+    backend: str = field(init=False, default='pt')
+    def get_activation(self, name: str) -> Any:
+        """Store copy of hidden unit activations at each layer of model."""
+
+        def hook(model, input, output):
+            # store copy of tensor rather than tensor itself
+            if isinstance(output, tuple):
+                act = output[0]
+            else:
+                act = output
+            try:
+                activations[name] = act.clone().detach()
+            except AttributeError:
+                activations[name] = act.clone()
+
+        return hook
+
+    def register_hook(self) -> Any:
+        """Register a forward hook to store activations."""
+        for n, m in self.model.named_modules():
+            m.register_forward_hook(self.get_activation(n))
+
+    @torch.no_grad()
+    def _extract_features(
+        self,
+        batches: Iterator,
+        module_name: str,
+        flatten_acts: bool
+    ) -> Array:
+        device = torch.device(self.device)
+        self.model = self.model.to(device)
+        # initialise an empty dict to store features for each mini-batch
+        global activations
+        activations = {}
+        # register forward hook to store features
+        self.register_hook()
+        features = []
+        for batch in tqdm(batches, desc="Batch"):
+            batch = batch.to(device)
+            _ = self.forward(batch)
+            act = activations[module_name]
+            if flatten_acts:
+                act = self.flatten_acts(act)
+            features.append(act.cpu().numpy())
+        features = np.vstack(features)
+        return features
+
+    def forward(self, batch: Tensor) -> Tensor:
+        """Default forward pass."""
+        if(self.model_name == 'OpenCLIP'):
```

Comment:
This should probably be outsourced into a separate forward function for that specific custom model @RLeipe @LukasMut

Reply:
@andropar Yes, I agree. That makes sense to me. Do you want to change it in the current PR?

Reply:
Will do 👍